### DISCUSSION TITLE
Hardware requirements

### DISCUSSION THREAD

**Discussion by onahman** [BayraktarLab/cell2location#16] [Q&A] [answered]
Hi,
I have ran the first part of the tutorial using 1 CPU and it took REALLY long time (almost 11 hours).
Can you please write down what hardware you used or alternatively minimal hardware requirements?
I am looking for a better computing solution and this might be a great help

Ornit

**Comment by vitkl** [comment, marked as answer]
Hi @onahman 

There are 2 parts to this: 1) general and 2) your specific issue.

1) A. - We generally use GPU for both step 1 and 2 of the tutorial. Any GPU with sufficient memory for your dataset would work but the fastest results are with Tesla V100 (< 5 minutes for step 1, 25 min for step 2). The tutorials need about 2-5 GB of GPU memory.
B. - RAM memory requirements are more stringent. To perform the second step, for every 1GB of GPU memory expect 10 GB of RAM (10GB of GPU -> 100GB of RAM).
C. - We also provide an experimental numpyro translation of the model which has improved memory efficiency (allowing analysis of multiple Visium samples on Google Colab) and minor improvements in speed - https://github.com/vitkl/cell2location_numpyro. It can also work in several hours on CPU. You can try it on Google Colab [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/vitkl/cell2location_numpyro/blob/main/docs/notebooks/cell2location_short_demo_colab.ipynb) - however note that both numpyro itself and cell2location_numpyro are in very active development.
D. - Computations needed for NB distribution in theano are implemented in a way which makes running on CPU and even parallelisation over CPU slow. So GPU is necessary except for small datasets (10s - 100 of samples).

2) 11 hour run time should not be the case for the first step. Running on 1 CPU should be around 17 times slower than GPU - 4h total run time for the demo notebook, [cell2location_estimating_signatures-cpu.ipynb.zip](https://github.com/BayraktarLab/cell2location/files/5754379/cell2location_estimating_signatures-cpu.ipynb.zip)). Could you please provide more details on how you are running it? 
You can speed up run time if you increase to > 4 available CPUs like this at the beginning of the notebook:
```python
import torch
torch.set_num_threads(8)
```

If you are running it on Mac, slow speed might be due to insufficient RAM. You can also try reducing the initial number of epochs to 20 or 50. Finally, you can try it on Google Colab with GPU.
